torch/csrc/jit/pybind_utils.h

```diff
@@ -22,52 +26,143 @@ inline Stack createStack(const py::tuple& tuple, at::ArrayRef<Value*> inputs, si
+    const py::object& object) {
+  try {
+    return toIValue(object, argument.type);
+  } catch (const py::cast_error& error) {
+    AT_ERROR(
+        "Expected value of type ", *argument.type,
+        " for argument '", argument.name,
+        "' in position ", argumentPosition,
+        ", but instead got value of type ", object.get_type().str());
+  }
+}
+
+inline py::object toPyObject(const IValue& ivalue) {
+  switch (ivalue.getTag()) {
+    case IValue::Tag::None:
+      return py::none();
+    case IValue::Tag::Tensor:
+      return py::cast(autograd::Variable(ivalue.toTensor()));
+    case IValue::Tag::Double:
+      return py::cast(ivalue.toDouble());
+    case IValue::Tag::Int:
+      return py::cast(ivalue.toInt());
+    case IValue::Tag::Tuple:
+    case IValue::Tag::IntList:
+    case IValue::Tag::DoubleList:
+    case IValue::Tag::TensorList:
+      AT_ERROR("Lists and tuples are not yet supported");
+    default:
+      AT_ERROR("Missing cases in toPyObject! File a bug report.");
+  }
+}
+
+inline void findErrorInKwargs(
+    const std::vector<Argument>& arguments,
+    py::kwargs kwargs) {
+  for (const auto& kwarg : kwargs) {
+    const auto key = py::cast<std::string>(kwarg.first);
+    AT_CHECK(
+        std::count_if(
+            arguments.begin(),
+            arguments.end(),
+            [&key](const Argument& argument) { return argument.name == key; }),
+        "Unknown keyword argument '", key, "'");
+  }
+  // If there are unconsumed kwargs but none of them were uknown, the first
+  // positional argument present in the kwargs is duplicated.
+  for (const auto& argument : arguments) {
+    if (kwargs.contains(argument.name.c_str())) {
+      AT_ASSERT(!argument.default_value);
+      AT_ERROR(
+          "Argument '",
+          argument.name,
+          "' specified both as positional and keyword argument");
     }
-    throw std::runtime_error("Missing cases in castToIValue! File a bug report.");
-  };
+  }
+}
+
+inline Stack createStackForSchema(
+    const FunctionSchema& schema,
+    py::args args,
+    py::kwargs kwargs) {
+  AT_CHECK(
+      args.size() + kwargs.size() <= schema.arguments.size(),
+      "Expected at most ", schema.arguments.size(),
+      " arguments for operator ", schema.name, ", but received ",
+      args.size(), " arguments");
+
+  Stack stack;
+  stack.reserve(args.size() + kwargs.size());
+
+  // First push all positional args.
+  for (size_t i = 0; i < args.size(); ++i) {
+    // Use the type information from the schema to convert the PyObject.
+    push(stack, argumentToIValue(i, schema.arguments[i], args[i]));
+  }
+
+  // Now for every remaining non-positional argument in the schema, look for it
+  // in the kwargs dict and push it if found, or use its default value if it
+  // has one.
+  size_t consumed_kwargs = 0;
+  for (size_t i = args.size(); i < schema.arguments.size(); ++i) {
+    const auto& arg = schema.arguments[i];
+    if (kwargs.contains(arg.name.c_str())) {
```

Comment:
Is there a way to avoid the double-hashing here?

Reply:
Looked, but can't see any convenient way in the pybind API ...